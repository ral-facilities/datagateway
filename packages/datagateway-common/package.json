{
  "name": "datagateway-common",
  "version": "0.1.0",
  "private": true,
  "files": [
    "lib"
  ],
  "main": "./lib/index.js",
  "dependencies": {
    "@date-io/date-fns": "^1.3.13",
    "@material-ui/pickers": "^3.2.10",
    "@material-ui/lab": "^4.0.0-alpha.57",
    "@types/lodash.debounce": "^4.0.6",
    "axios": "^0.21.1",
    "clsx": "^1.1.1",
    "connected-react-router": "^6.9.1",
    "date-fns": "^2.17.0",
    "i18next": "^19.9.0",
    "lodash.debounce": "^4.0.8",
    "loglevel": "^1.7.1",
    "react-draggable": "^4.4.3",
    "react-i18next": "^11.8.8",
    "react-redux": "^7.1.0",
    "react-router": "^5.0.1",
    "react-scripts": "4.0.3",
    "react-virtualized": "^9.22.3",
    "redux": "^4.0.4",
    "redux-mock-store": "^1.5.4",
    "redux-thunk": "^2.3.0"
  },
  "peerDependencies": {
    "@material-ui/core": ">=  4.0.0 < 5",
    "@material-ui/icons": ">=  4.0.0 < 5",
    "react": ">= 16.8.0 < 17",
    "react-dom": ">= 16.8.0 < 17",
    "react-router-dom": ">= 5.0.0 < 6"
  },
  "devDependencies": {
    "@material-ui/core": "^4.11.3",
    "@material-ui/icons": "^4.11.2",
    "@types/enzyme": "^3.10.8",
    "@types/jest": "^26.0.4",
    "@types/node": "^14.14.31",
    "@types/react": "^17.0.2",
    "@types/react-router-dom": "^5.1.7",
    "@types/react-virtualized": "^9.21.10",
    "@typescript-eslint/eslint-plugin": "^4.16.1",
    "@typescript-eslint/parser": "^4.16.1",
    "babel-eslint": "10.1.0",
    "enzyme": "^3.10.0",
    "enzyme-adapter-react-16": "^1.15.6",
    "enzyme-to-json": "^3.6.1",
    "eslint": "^7.5.0",
    "eslint-config-prettier": "^8.1.0",
    "eslint-config-react-app": "^6.0.0",
    "eslint-plugin-cypress": "^2.11.1",
    "eslint-plugin-flowtype": "^5.8.0",
    "eslint-plugin-import": "^2.22.1",
    "eslint-plugin-jsx-a11y": "^6.3.1",
    "eslint-plugin-prettier": "^3.3.1",
    "eslint-plugin-react": "^7.20.3",
    "eslint-plugin-react-hooks": "^4.1.0",
    "lint-staged": "^10.5.4",
    "prettier": "^2.2.1",
    "react": "^16.13.1",
    "react-dom": "^16.11.0",
    "react-router-dom": "^5.0.1",
    "typescript": "4.2.2",
    "tslib": "^2.1.0"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test --env=jsdom --coverage --watchAll=false",
    "prepare": "yarn tsc",
    "test:watch": "react-scripts test --env=jsdom --watch",
    "lint:js": "eslint --ext=tsx --ext=ts --ext=js --ext=jsx --fix ./src",
    "tsc": "tsc -p tsconfig.lib.json",
    "eject": "react-scripts eject",
    "pre-commit": "lint-staged"
  },
  "lint-staged": {
    "src/**/*.{tsx,ts,js,jsx,json}": [
      "eslint --max-warnings=0 --ext=tsx --ext=ts --ext=js --ext=jsx  --fix",
      "prettier --write"
    ],
    "cypress/**/*.{tsx,ts,js,jsx,json}": [
      "eslint --max-warnings=0 --ext=tsx --ext=ts --ext=js --ext=jsx --fix",
      "prettier --write"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ]
  },
  "jest": {
    "snapshotSerializers": [
      "enzyme-to-json/serializer"
    ],
    "collectCoverageFrom": [
      "src/**/*.{tsx,ts,js,jsx}",
      "!src/index.tsx",
      "!src/serviceWorker.ts",
      "!src/setupTests.js"
    ]
  }
}
