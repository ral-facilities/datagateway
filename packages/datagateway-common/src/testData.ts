import type {
  Datafile,
  Dataset,
  FacilityCycle,
  Investigation,
} from './app.types';

export const mockApiUrl = 'https://example.com/api';
export const mockFacilityName = 'LILS';

export const mockFacilityCycles: FacilityCycle[] = [
  {
    id: 12938,
    name: 'studio toughness',
    description: 'He decided water-skiing on a frozen lake wasnâ€™t a good idea.',
    startDate: '2006-01-20T16:30:17Z',
    endDate: '2007-01-20T16:30:17Z',
  },
  {
    id: 402,
    name: 'within cell interlinked',
    description: 'He waited for the stop sign to turn to a go sign.',
    startDate: '2017-03-17T14:03:11Z',
    endDate: '2020-11-29T05:41:54Z',
  },
];

export const mockInvestigations: Investigation[] = [
  {
    id: 58,
    title: 'Happiness can be found in the depths of chocolate pudding.',
    name: 'investigation news',
    visitId: 'CqJN',
    startDate: '2018-03-09T08:19:55Z',
    endDate: '2018-03-29T08:19:55Z',
    investigationInstruments: [
      {
        id: 446,
        instrument: {
          id: 937,
          name: 'instrument fame',
        },
      },
    ],
    investigationFacilityCycles: [
      {
        id: 23,
        facilityCycle: {
          id: 402,
          name: 'within cell interlinked',
          description: 'He waited for the stop sign to turn to a go sign.',
          startDate: '2017-03-17T14:03:11Z',
          endDate: '2020-11-29T05:41:54Z',
        },
      },
    ],
  },
  {
    id: 993,
    title: 'I ate a sock because people on the Internet told me to.',
    name: 'investigation inn',
    visitId: 'z0bLi1f3',
    startDate: '2019-08-08T22:27:07Z',
    endDate: '2019-09-08T22:27:07Z',
    investigationInstruments: [
      {
        id: 262,
        instrument: {
          id: 927,
          name: 'instrument case',
        },
      },
    ],
    investigationFacilityCycles: [],
  },
];

export const mockDatasets: Dataset[] = [
  {
    id: 856,
    name: 'dataset modern',
    investigation: mockInvestigations[0],
    createTime: '2018-03-10T08:19:55Z',
    modTime: '2018-03-10T09:19:55Z',
  },
  {
    id: 535,
    name: 'dataset lazy',
    investigation: mockInvestigations[1],
    createTime: '2019-09-01T22:27:07Z',
    modTime: '2019-09-02T22:27:07Z',
  },
];

export const mockDatafiles: Datafile[] = [
  {
    id: 70,
    name: 'datafile weekend',
    modTime: '2018-03-10T08:19:55Z',
    createTime: '2018-03-10T08:19:55Z',
    dataset: mockDatasets[0],
  },
];
