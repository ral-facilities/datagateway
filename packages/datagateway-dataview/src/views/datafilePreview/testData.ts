// This file contains mock data for testing datafile preview functionality.

import type { Datafile } from 'datagateway-common';

const mockDatafile: Datafile = {
  createTime: '2020-01-01',
  dataset: undefined,
  description: 'test description',
  fileSize: 100,
  id: 123,
  location: 'test location',
  modTime: '2020-01-02',
  name: 'Datafile.txt',
  parameters: [],
};

const mockTxtFileContent = `First line
Second line
Third line
`;

export { mockDatafile, mockTxtFileContent };
