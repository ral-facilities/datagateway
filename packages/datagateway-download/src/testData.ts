import {
  Download,
  DownloadCartItem,
  FormattedDownload,
} from 'datagateway-common';
import type { DownloadSettings } from './ConfigProvider';

export const mockDownloadItems: Download[] = [
  {
    createdAt: '2020-02-25T15:05:29Z',
    downloadItems: [{ entityId: 1, entityType: 'investigation', id: 1 }],
    email: 'test1@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-1',
    fullName: 'Person 1',
    id: 1,
    isDeleted: false,
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 1000,
    status: 'COMPLETE',
    transport: 'https',
    userName: 'test user',
  },
  {
    createdAt: '2020-02-26T15:05:35Z',
    downloadItems: [{ entityId: 2, entityType: 'investigation', id: 2 }],
    email: 'test2@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-2',
    fullName: 'Person 2',
    id: 2,
    isDeleted: false,
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 2000,
    status: 'PREPARING',
    transport: 'globus',
    userName: 'test user',
  },
  {
    createdAt: '2020-02-27T15:57:20Z',
    downloadItems: [{ entityId: 3, entityType: 'investigation', id: 3 }],
    email: 'test3@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-3',
    fullName: 'Person 3',
    id: 3,
    isDeleted: false,
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 3000,
    status: 'RESTORING',
    transport: 'https',
    userName: 'test user',
  },
  {
    createdAt: '2020-02-28T15:57:28Z',
    downloadItems: [{ entityId: 4, entityType: 'investigation', id: 4 }],
    email: 'test4@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-4',
    fullName: 'Person 4',
    id: 4,
    isDeleted: true,
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 4000,
    status: 'EXPIRED',
    transport: 'globus',
    userName: 'test user',
  },
  {
    createdAt: '2020-03-01T15:57:28Z[UTC]',
    downloadItems: [{ entityId: 5, entityType: 'investigation', id: 5 }],
    email: 'test5@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-5',
    fullName: 'Person 5',
    id: 5,
    isDeleted: false,
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 5000,
    status: 'PAUSED',
    transport: 'globus',
    userName: 'test user',
  },
];

export const mockFormattedDownloadItems: FormattedDownload[] = [
  {
    createdAt: '2020-02-25T15:05:29Z',
    downloadItems: [{ entityId: 1, entityType: 'investigation', id: 1 }],
    email: 'test1@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-1',
    fullName: 'Person 1',
    id: 1,
    isDeleted: false,
    formattedIsDeleted: 'No',
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 1000,
    status: 'COMPLETE',
    formattedStatus: 'downloadStatus.complete',
    transport: 'https',
    userName: 'test user',
  },
  {
    createdAt: '2020-02-26T15:05:35Z',
    downloadItems: [{ entityId: 2, entityType: 'investigation', id: 2 }],
    email: 'test2@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-2',
    fullName: 'Person 2',
    id: 2,
    isDeleted: false,
    formattedIsDeleted: 'No',
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 2000,
    status: 'PREPARING',
    formattedStatus: 'downloadStatus.preparing',
    transport: 'globus',
    userName: 'test user',
  },
  {
    createdAt: '2020-02-27T15:57:20Z',
    downloadItems: [{ entityId: 3, entityType: 'investigation', id: 3 }],
    email: 'test3@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-3',
    fullName: 'Person 3',
    id: 3,
    isDeleted: false,
    formattedIsDeleted: 'No',
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 3000,
    status: 'RESTORING',
    formattedStatus: 'downloadStatus.restoring',
    transport: 'https',
    userName: 'test user',
  },
  {
    createdAt: '2020-02-28T15:57:28Z',
    downloadItems: [{ entityId: 4, entityType: 'investigation', id: 4 }],
    email: 'test4@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-4',
    fullName: 'Person 4',
    id: 4,
    isDeleted: true,
    formattedIsDeleted: 'Yes',
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 4000,
    status: 'EXPIRED',
    formattedStatus: 'downloadStatus.expired',
    transport: 'globus',
    userName: 'test user',
  },
  {
    createdAt: '2020-03-01T15:57:28Z[UTC]',
    downloadItems: [{ entityId: 5, entityType: 'investigation', id: 5 }],
    email: 'test5@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-5',
    fullName: 'Person 5',
    id: 5,
    isDeleted: false,
    formattedIsDeleted: 'No',
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 5000,
    status: 'PAUSED',
    formattedStatus: 'downloadStatus.paused',
    transport: 'globus',
    userName: 'test user',
  },
];

export const mockCartItems: DownloadCartItem[] = [
  {
    entityId: 1,
    entityType: 'investigation',
    id: 1,
    name: 'INVESTIGATION 1',
    parentEntities: [],
  },
  {
    entityId: 2,
    entityType: 'investigation',
    id: 2,
    name: 'INVESTIGATION 2',
    parentEntities: [],
  },
  {
    entityId: 3,
    entityType: 'dataset',
    id: 3,
    name: 'DATASET 1',
    parentEntities: [],
  },
  {
    entityId: 4,
    entityType: 'datafile',
    id: 4,
    name: 'DATAFILE 1',
    parentEntities: [],
  },
];

// Create our mocked datagateway-download settings file.
export const mockedSettings: Partial<DownloadSettings> = {
  facilityName: 'LILS',
  apiUrl: 'https://example.com/api',
  downloadApiUrl: 'https://example.com/downloadApi',
  idsUrl: 'https://example.com/ids',
  fileCountMax: 5000,
  totalSizeMax: 1000000000000,
  accessMethods: {
    https: {
      idsUrl: 'https://example.com/ids',
      displayName: 'HTTPS',
      description: 'Example description for <b>HTTPS</b> access method.',
    },
    globus: {
      idsUrl: 'https://example.com/ids',
      displayName: 'Globus',
      description: 'Example description for Globus access method.',
    },
  },
  uiFeatures: {
    downloadProgress: false,
  },
};
