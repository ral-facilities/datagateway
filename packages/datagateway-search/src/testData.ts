// Defines test data used for testing

import type { Investigation } from 'datagateway-common';

const mockInvestigation: Investigation = {
  id: 1,
  name: 'Test investigation name',
  title: 'Test investigation title',
  visitId: '1',
};

export { mockInvestigation };
