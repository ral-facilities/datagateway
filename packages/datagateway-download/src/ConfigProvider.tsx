import { PluginRoute, Preloader } from 'datagateway-common';
import React from 'react';
import { settings } from './settings';

export interface DownloadSettingsAccessMethod {
  [type: string]: {
    idsUrl: string;
    displayName?: string;
    description?: string;
  };
}

export interface DownloadSettings {
  facilityName: string;
  apiUrl: string;
  downloadApiUrl: string;
  idsUrl: string;

  fileCountMax?: number;
  totalSizeMax?: number;

  accessMethods: DownloadSettingsAccessMethod;
  routes: PluginRoute[];
  helpSteps: { target: string; content: string }[];
  pluginHost?: string;

  /**
   * A map of UI flags that can toggle certain UI features..
   */
  uiFeatures: {
    /**
     * Enables download progress to be displayed in download status tables when set to true.
     */
    downloadProgress: boolean;
  };
}

const initialConfiguration: DownloadSettings = {
  facilityName: '',
  apiUrl: '',
  downloadApiUrl: '',
  idsUrl: '',
  fileCountMax: undefined,
  totalSizeMax: undefined,
  accessMethods: {},
  routes: [],
  helpSteps: [],
  uiFeatures: {
    downloadProgress: false,
  },
};

export const DownloadSettingsContext =
  React.createContext<DownloadSettings>(initialConfiguration);

class ConfigProvider extends React.Component<
  { children: React.ReactNode },
  { loading: boolean; settings: DownloadSettings }
> {
  public constructor(props: { children: React.ReactNode }) {
    super(props);

    this.state = {
      loading: true,
      settings: initialConfiguration,
    };
  }

  public componentDidMount(): void {
    this.updateConfigurationState();
  }

  private updateConfigurationState = async (): Promise<void> => {
    const settingsResult = await settings;
    if (settingsResult) {
      this.setState({
        loading: false,
        settings: settingsResult,
      });
    }
  };

  public render(): React.ReactElement {
    return (
      // We pass the download settings that has been loaded
      // for all child components to consume, if required.
      <Preloader loading={this.state.loading}>
        <DownloadSettingsContext.Provider value={this.state.settings}>
          {this.props.children}
        </DownloadSettingsContext.Provider>
      </Preloader>
    );
  }
}

export const ConfigConsumer = DownloadSettingsContext.Consumer;
export default ConfigProvider;
