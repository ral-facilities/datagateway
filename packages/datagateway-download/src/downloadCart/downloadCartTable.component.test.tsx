import {
  fireEvent,
  render,
  RenderResult,
  screen,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import type { UserEvent } from '@testing-library/user-event/dist/types/setup';
import { fetchDownloadCart } from 'datagateway-common';
import { createMemoryHistory } from 'history';
import * as React from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { Router } from 'react-router-dom';
import { DownloadSettingsContext } from '../ConfigProvider';
import { mockCartItems, mockedSettings } from '../testData';
import {
  getDatafileCount,
  getSize,
  removeAllDownloadCartItems,
  removeFromCart,
} from '../downloadApi';
import DownloadCartTable from './downloadCartTable.component';

jest.mock('datagateway-common', () => {
  const originalModule = jest.requireActual('datagateway-common');

  return {
    __esModule: true,
    ...originalModule,
    fetchDownloadCart: jest.fn(),
  };
});

jest.mock('../downloadApi', () => {
  const originalModule = jest.requireActual('../downloadApi');

  return {
    ...originalModule,
    removeAllDownloadCartItems: jest.fn(),
    getSize: jest.fn(),
    getDatafileCount: jest.fn(),
    getIsTwoLevel: jest.fn().mockResolvedValue(true),
    removeFromCart: jest.fn(),
  };
});

const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

const renderComponent = (): RenderResult =>
  render(
    <QueryClientProvider client={createTestQueryClient()}>
      <DownloadSettingsContext.Provider value={mockedSettings}>
        <Router history={createMemoryHistory()}>
          <DownloadCartTable statusTabRedirect={jest.fn()} />
        </Router>
      </DownloadSettingsContext.Provider>
    </QueryClientProvider>
  );

describe('Download cart table component', () => {
  let holder, queryClient;
  let user: UserEvent;

  const resetDOM = (): void => {
    if (holder) document.body.removeChild(holder);
    holder = document.getElementById('datagateway-download');
    if (holder) document.body.removeChild(holder);
  };

  beforeEach(() => {
    user = userEvent.setup();
    queryClient = new QueryClient();

    //https://stackoverflow.com/questions/43694975/jest-enzyme-using-mount-document-getelementbyid-returns-null-on-componen
    holder = document.createElement('div');
    holder.setAttribute('id', 'datagateway-download');
    document.body.appendChild(holder);

    (
      fetchDownloadCart as jest.MockedFunction<typeof fetchDownloadCart>
    ).mockResolvedValue(mockCartItems);
    (
      removeAllDownloadCartItems as jest.MockedFunction<
        typeof removeAllDownloadCartItems
      >
    ).mockResolvedValue(undefined);
    (
      removeFromCart as jest.MockedFunction<typeof removeFromCart>
    ).mockImplementation((entityType, entityIds) => {
      return Promise.resolve(
        mockCartItems.filter((item) => !entityIds.includes(item.entityId))
      );
    });

    (getSize as jest.MockedFunction<typeof getSize>).mockResolvedValue(1);
    (
      getDatafileCount as jest.MockedFunction<typeof getDatafileCount>
    ).mockResolvedValue(7);
  });

  afterEach(() => {
    resetDOM();
    jest.clearAllMocks();
    jest.clearAllTimers();
    jest.useRealTimers();
  });

  it('should render no cart message correctly', async () => {
    (fetchDownloadCart as jest.Mock).mockResolvedValue([]);

    renderComponent();

    expect(
      await screen.findByText('No data selected', { exact: false })
    ).toBeTruthy();
  });

  it('should show download sizes for cart items', async () => {
    renderComponent();

    expect(
      await screen.findByText('downloadCart.total_size: 4 B / 1 TB')
    ).toBeTruthy();
  });

  it('should show progress indicator when calculating file count of cart', async () => {
    (getDatafileCount as jest.Mock).mockImplementation(
      () =>
        new Promise((_) => {
          // never resolve promise so that progress indicator stays visible.
        })
    );

    jest.useFakeTimers();

    renderComponent();

    expect(
      await screen.findByLabelText('downloadCart.calculating')
    ).toBeInTheDocument();
  });

  it('should show total file count of the cart', async () => {
    renderComponent();

    expect(
      await screen.findByText('downloadCart.number_of_files: 22 / 5000')
    ).toBeTruthy();
  });

  it('should load cart confirmation dialog when Download Cart button is clicked', async () => {
    renderComponent();

    await user.click(await screen.findByText('downloadCart.download'));

    expect(
      await screen.findByLabelText('downloadConfirmDialog.dialog_arialabel')
    ).toBeTruthy();
  });

  it('should remove all items from cart when Remove All button is clicked', async () => {
    renderComponent();

    await user.click(await screen.findByText('downloadCart.remove_all'));

    expect(
      await screen.findByText('No data selected', { exact: false })
    ).toBeTruthy();
  });

  it('disables remove all button while request is processing', async () => {
    // use this to manually resolve promise
    let promiseResolve;

    (
      removeAllDownloadCartItems as jest.MockedFunction<
        typeof removeAllDownloadCartItems
      >
    ).mockImplementation(
      () =>
        new Promise((resolve) => {
          promiseResolve = resolve;
        })
    );

    renderComponent();

    await user.click(
      await screen.findByRole('button', { name: 'downloadCart.remove_all' })
    );

    expect(
      await screen.findByRole('button', { name: 'downloadCart.remove_all' })
    ).toBeDisabled();

    promiseResolve();

    expect(
      await screen.findByTestId('no-selections-message')
    ).toBeInTheDocument();
  });

  it('should disable download button when there are empty items in the cart ', async () => {
    (getSize as jest.MockedFunction<typeof getSize>).mockResolvedValueOnce(0);
    (
      getDatafileCount as jest.MockedFunction<typeof getDatafileCount>
    ).mockResolvedValueOnce(0);

    renderComponent();

    expect(
      await screen.findByText('downloadCart.empty_items_error')
    ).toBeTruthy();

    const downloadButton = await screen.findByText('downloadCart.download');
    // cannot use user.click here because clicking disabled button will throw
    fireEvent.click(downloadButton);
    // should not show confirm dialog
    expect(
      screen.queryByText('downloadConfirmDialog.dialog_arialabel')
    ).toBeNull();
  });

  it("should remove an item when said item's remove button is clicked", async () => {
    renderComponent();

    await user.click(
      await screen.findByLabelText('downloadCart.remove {name:INVESTIGATION 2}')
    );

    await waitFor(async () => {
      expect(screen.queryByText('INVESTIGATION 2')).toBeNull();
    });
  });

  it('should sort data when headers are clicked', async () => {
    renderComponent();

    const typeSortLabel = await screen.findByRole('button', {
      name: 'downloadCart.type',
    });

    await user.click(typeSortLabel);

    let rows = await screen.findAllByText(
      /(DATAFILE|DATASET|INVESTIGATION) \d/
    );
    // row should be sorted by type asc.
    expect(rows[0]).toHaveTextContent('DATAFILE 1');
    expect(rows[1]).toHaveTextContent('DATASET 1');
    expect(rows[2]).toHaveTextContent('INVESTIGATION 1');
    expect(rows[3]).toHaveTextContent('INVESTIGATION 2');

    await user.click(typeSortLabel);

    rows = await screen.findAllByText(/(DATAFILE|DATASET|INVESTIGATION) \d/);
    // row should be sorted by type desc.
    expect(rows[0]).toHaveTextContent('INVESTIGATION 1');
    expect(rows[1]).toHaveTextContent('INVESTIGATION 2');
    expect(rows[2]).toHaveTextContent('DATASET 1');
    expect(rows[3]).toHaveTextContent('DATAFILE 1');

    const nameSortLabel = await screen.findByRole('button', {
      name: 'downloadCart.name',
    });

    await user.click(nameSortLabel);

    rows = await screen.findAllByText(/(DATAFILE|DATASET|INVESTIGATION) \d/);
    // row should be sorted by type desc & name asc.
    expect(rows[0]).toHaveTextContent('INVESTIGATION 1');
    expect(rows[1]).toHaveTextContent('INVESTIGATION 2');
    expect(rows[2]).toHaveTextContent('DATASET 1');
    expect(rows[3]).toHaveTextContent('DATAFILE 1');

    await user.click(nameSortLabel);

    rows = await screen.findAllByText(/(DATAFILE|DATASET|INVESTIGATION) \d/);
    // row should be sorted by type desc & name desc.
    expect(rows[0]).toHaveTextContent('INVESTIGATION 2');
    expect(rows[1]).toHaveTextContent('INVESTIGATION 1');
    expect(rows[2]).toHaveTextContent('DATASET 1');
    expect(rows[3]).toHaveTextContent('DATAFILE 1');

    await user.click(nameSortLabel);

    rows = await screen.findAllByText(/(DATAFILE|DATASET|INVESTIGATION) \d/);
    // row should be sorted by type desc.
    expect(rows[0]).toHaveTextContent('INVESTIGATION 1');
    expect(rows[1]).toHaveTextContent('INVESTIGATION 2');
    expect(rows[2]).toHaveTextContent('DATASET 1');
    expect(rows[3]).toHaveTextContent('DATAFILE 1');
  }, 10000);

  it('should filter data when text fields are typed into', async () => {
    renderComponent();

    // TEST NAME FILTER INPUT

    expect(await screen.findByText('INVESTIGATION 2')).toBeTruthy();

    const nameFilterInput = await screen.findByLabelText(
      'Filter by downloadCart.name'
    );
    await user.type(nameFilterInput, '1');

    await waitFor(async () => {
      expect(screen.queryByText('INVESTIGATION 2')).toBeNull();
    });

    await user.clear(nameFilterInput);

    expect(await screen.findByText('INVESTIGATION 2')).toBeInTheDocument();

    // TEST TYPE FILTER INPUT

    expect(await screen.findByText('dataset')).toBeTruthy();
    expect(await screen.findByText('datafile')).toBeTruthy();

    const typeFilterInput = await screen.findByLabelText(
      'Filter by downloadCart.type'
    );
    await user.type(typeFilterInput, 'investigation');

    await waitFor(async () => {
      expect(screen.queryByText('dataset')).toBeNull();
      expect(screen.queryByText('datafile')).toBeNull();
    });
  });

  it('should display error alert if file/size limit exceeded', async () => {
    renderComponent();

    await waitFor(() => {
      expect(
        screen.queryByText('downloadCart.file_limit_error', { exact: false })
      ).toBeNull();
      expect(
        screen.queryByText('downloadCart.size_limit_error', { exact: false })
      ).toBeNull();
    });

    resetDOM();
    render(
      <QueryClientProvider client={queryClient}>
        <DownloadSettingsContext.Provider
          value={{
            ...mockedSettings,
            totalSizeMax: 1,
          }}
        >
          <Router history={createMemoryHistory()}>
            <DownloadCartTable statusTabRedirect={jest.fn()} />
          </Router>
        </DownloadSettingsContext.Provider>
      </QueryClientProvider>
    );

    expect(
      await screen.findByText(
        'downloadCart.size_limit_error {totalSizeMax:1 B}',
        { exact: false }
      )
    ).toBeTruthy();

    resetDOM();
    render(
      <QueryClientProvider client={queryClient}>
        <DownloadSettingsContext.Provider
          value={{
            ...mockedSettings,
            fileCountMax: 1,
          }}
        >
          <Router history={createMemoryHistory()}>
            <DownloadCartTable statusTabRedirect={jest.fn()} />
          </Router>
        </DownloadSettingsContext.Provider>
      </QueryClientProvider>
    );

    expect(
      await screen.findByText(
        'downloadCart.file_limit_error {fileCountMax:1}',
        { exact: false }
      )
    ).toBeTruthy();
  });

  it('does not display error alerts if file/size limits are not set', async () => {
    render(
      <QueryClientProvider client={queryClient}>
        <DownloadSettingsContext.Provider
          value={{
            ...mockedSettings,
            fileCountMax: undefined,
            totalSizeMax: undefined,
          }}
        >
          <Router history={createMemoryHistory()}>
            <DownloadCartTable statusTabRedirect={jest.fn()} />
          </Router>
        </DownloadSettingsContext.Provider>
      </QueryClientProvider>
    );

    await waitFor(() => {
      expect(
        screen.queryByText('downloadCart.file_limit_error', { exact: false })
      ).toBeNull();
      expect(
        screen.queryByText('downloadCart.size_limit_error', { exact: false })
      ).toBeNull();
    });

    expect(
      await screen.findByText('downloadCart.total_size: 4 B', { exact: true })
    ).toBeTruthy();
    expect(
      await screen.findByText('downloadCart.number_of_files: 22', {
        exact: true,
      })
    ).toBeTruthy();
  });
});
