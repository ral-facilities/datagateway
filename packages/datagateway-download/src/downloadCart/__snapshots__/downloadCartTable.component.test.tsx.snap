// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Download cart table component renders correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Grid))
    container={true}
    direction="column"
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={12}
    >
      <WithStyles(ForwardRef(LinearProgress))
        color="secondary"
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
    >
      <WithStyles(ForwardRef(Paper))
        className="tour-download-results"
        style={
          Object {
            "height": "calc(100vh - 64px - 30px - 48px - 48px - 3rem - (1.75 * 0.875rem + 12px)",
            "minHeight": 230,
            "overflowX": "auto",
          }
        }
      >
        <WithStyles(VirtualizedTable)
          actions={
            Array [
              [Function],
            ]
          }
          columns={
            Array [
              Object {
                "dataKey": "name",
                "filterComponent": [Function],
                "label": "downloadCart.name",
              },
              Object {
                "dataKey": "entityType",
                "filterComponent": [Function],
                "label": "downloadCart.type",
              },
              Object {
                "cellContentRenderer": [Function],
                "dataKey": "size",
                "label": "downloadCart.size",
              },
            ]
          }
          data={Array []}
          loading={true}
          onSort={[Function]}
          sort={Object {}}
        />
      </WithStyles(ForwardRef(Paper))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      alignItems="flex-end"
      container={true}
      direction="column"
      item={true}
      justify="space-between"
    >
      <WithStyles(ForwardRef(Grid))
        alignContent="flex-end"
        container={true}
        direction="column"
        item={true}
        style={
          Object {
            "marginRight": "1.2em",
          }
        }
        xs={true}
      >
        <WithStyles(ForwardRef(Typography))
          id="fileCountDisplay"
        >
          downloadCart.number_of_files
          :
           
          0
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          id="totalSizeDisplay"
        >
          downloadCart.total_size
          :
           
          0 B
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        container={true}
        item={true}
        justify="flex-end"
        spacing={1}
        style={
          Object {
            "marginRight": "1em",
          }
        }
        xs={true}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Button))
            className="tour-download-remove-button"
            color="primary"
            id="removeAllButton"
            onClick={[Function]}
            variant="contained"
          >
            downloadCart.remove_all
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Button))
            className="tour-download-download-button"
            color="primary"
            disabled={true}
            id="downloadCartButton"
            onClick={[Function]}
            variant="contained"
          >
            downloadCart.download
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(DownloadConfirmDialog)
    aria-labelledby="downloadCartConfirmation"
    clearCart={[Function]}
    isTwoLevel={false}
    open={false}
    redirectToStatusTab={[MockFunction]}
    setClose={[Function]}
    totalSize={0}
  />
</div>
`;
